File: AndroidGameLibrary/src/cc/lib/android/DragAndDropAdapter.java
package cc.lib.android;

import android.content.ClipData;
import android.os.SystemClock;
import android.util.Log;
import android.view.DragEvent;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import cc.lib.game.Utils;

public abstract class DragAndDropAdapter<T> extends BaseAdapter implements View.OnDragListener, View.OnLongClickListener, View.OnTouchListener {

    final static String TAG = DragAndDropAdapter.class.getSimpleName();

    private final ListView dndLv;
    private final List<T> list = new ArrayList<>();
    private int dragInsertPos = -1;
    private int dragMinInsertPos = 0;
    private int dragMaxInsertPos = 0;
    private boolean dropped = false;

    public DragAndDropAdapter(ListView dndLv) {
        this.dndLv = dndLv;

    }

    public DragAndDropAdapter(ListView dndLv, List<T> items) {
        this(dndLv);
        list.addAll(items);
    }

    public DragAndDropAdapter(ListView dndLv, T ... items) {
        this(dndLv);
        list.addAll(Arrays.asList(items));
    }

    public final List<T> getList() {
        return list;
    }

    @Override
    public int getCount() {
        return list.size();
    }

    @Override
    public Object getItem(int position) {
        return list.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {
        if (convertView == null) {
            convertView = View.inflate(dndLv.getContext(), R.layout.draganddrop_list_item, null);
        }
        convertView.setOnDragListener(this);
        convertView.setOnLongClickListener(this);
        convertView.setTag(position);

        View divTop = convertView.findViewById(R.id.divider_top);
        View divBot = convertView.findViewById(R.id.divider_bottom);

        divTop.setVisibility(View.GONE);
        divBot.setVisibility(View.GONE);

        if (dragInsertPos >= 0) {
            if (dragInsertPos == position) {
                divTop.setVisibility(View.VISIBLE);
            } else if (dragInsertPos == position+1) {
                divBot.setVisibility(View.VISIBLE);
            }
        }

        final T cmd = list.get(position);

        TextView tvLineNum = (TextView) convertView.findViewById(R.id.tvLineNum);
        tvLineNum.setText(String.valueOf(position + 1));

        View v = convertView.findViewById(R.id.ibDelete);
        v.setVisibility(View.VISIBLE);
        v.setTag(position);
        v.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int position = (Integer) v.getTag();
                T cmd = list.remove(position);
                notifyDataSetChanged();
            }
        });


        ViewGroup container = (ViewGroup)convertView.findViewById(R.id.container);
        //container.removeAllViews();
        populateItem(cmd, container);

        return convertView;
    }

    protected abstract void populateItem(T cmd, ViewGroup container);

    @Override
    public boolean onDrag(View v, DragEvent event) {
        int dragInsertPosIn = dragInsertPos;
        String obj = null;
        Object [] state = (Object []) event.getLocalState();
        if (state == null)
            return false;
        T cmd = (T)state[0];
        int originatingLine = (Integer)state[1];
        if (v.equals(dndLv)) {
            obj = "ListView";
            switch (event.getAction()) {
                case DragEvent.ACTION_DROP: {
                    if (cmd != null) {
                        if (dragInsertPos < 0) {
                            list.add(cmd);
                        } else {
                            list.add(dragInsertPos, cmd);
                        }
                        notifyDataSetChanged();
                        dndLv.setSelection(dragInsertPos);
                        dragInsertPos = -1;
                    }
                    dropped = true;
                    break;
                }
                //case DragEvent.ACTION_DRAG_LOCATION:
                case DragEvent.ACTION_DRAG_ENTERED: {
                    if (list.size() > 0) {
                        dragInsertPos = Utils.clamp(list.size(), dragMinInsertPos, dragMaxInsertPos);
                        notifyDataSetChanged();
                    }
                    break;
                }
                case DragEvent.ACTION_DRAG_EXITED:
                    if (dragMinInsertPos < 0) {
                        dragInsertPos = -1;
                    } else if (originatingLine >= 0) {
                        dragInsertPos = originatingLine;
                    }
                    notifyDataSetChanged();
                    break;
                case DragEvent.ACTION_DRAG_STARTED:
                    dropped = false;
                    break;
                case DragEvent.ACTION_DRAG_ENDED:
                    dragInsertPos = -1;
                    if (!dropped && originatingLine >= 0) {
                        list.add(originatingLine, cmd);
                        notifyDataSetChanged();
                    }
                    break;
            }
        } else {
            int position = (Integer)v.getTag();
            obj = "RowItem[" + position + "]";
            switch (event.getAction()) {
                case DragEvent.ACTION_DROP: {
                    if (dragInsertPos < 0 || list.size() == 0) {
                        list.add(cmd);
                    } else {
                        list.add(dragInsertPos, cmd);
                    }
                    notifyDataSetChanged();
                    dndLv.setSelection(dragInsertPos);
                    dragInsertPos = -1;
                    dropped = true;
                    break;
                }
                case DragEvent.ACTION_DRAG_EXITED:
                    // TODO: Do I need this?
                    //dragInsertPos = -1;
                    //adapter.notifyDataSetChanged();
                    break;
                case DragEvent.ACTION_DRAG_LOCATION:
                case DragEvent.ACTION_DRAG_ENTERED: {
                    float mp = v.getHeight()/2;
                    float mpd = event.getY();// + dragging.getHeight();
                    int first = dndLv.getFirstVisiblePosition();
                    int last = dndLv.getLastVisiblePosition();

                    //if (position == first) {
                    //    Log.d("PLV", "A");
                    //    dragInsertPos = position;
                    //} else if (position == last) {
                    //    Log.d("PLV", "B");
                    //    dragInsertPos = position+1;
                    //} else
                    if (mp > mpd) {
                        Log.d(TAG, "C mp=" + mp + " mpd=" + mpd);
                        dragInsertPos = position;
                    } else {
                        Log.d(TAG, "D");
                        dragInsertPos = position+1;
                    }
                    dragInsertPos = Utils.clamp(dragInsertPos, dragMinInsertPos, dragMaxInsertPos);
                    notifyDataSetChanged();
                    //              setSelection(dragInsertPos);
                    if (dragInsertPosIn != dragInsertPos){
                        int scrollDist = v.getHeight();
                        int midpt =(last + first + 1) / 2;
                        float maxSp = 500f;
                        int speed = Math.round(maxSp - Math.abs(maxSp * 2 * (dragInsertPos-midpt)/((last-first))));
                        //0.5f * (dragInsertPos - first - midpt));
                        Log.d(TAG, "speed=" + speed);
                        if (dragInsertPos < midpt) {
                            dndLv.smoothScrollBy(-scrollDist, speed);
                        } else {
                            dndLv.smoothScrollBy(scrollDist, speed);
                        }
                    }

                    if (dragInsertPos == 0)
                        dndLv.smoothScrollToPositionFromTop(0, 0, 0);

                    if (dragInsertPos <= first+1) {
                        //    smoothScrollBy(-scrollDist, 500);
                    } else if (dragInsertPos >= last-1) {
                        //  smoothScrollBy(scrollDist, 500);
                    }
                    //            setSelection(dragInsertPos);
                    break;
                }
                case DragEvent.ACTION_DRAG_STARTED:

            }
        }
        //if (dragInsertPos != dragInsertPosIn)
        {
            Log.d(TAG, "v=" + obj + " action=" + getActionStr(event.getAction()) + " pos=" + dragInsertPos + " origLine: " + originatingLine);

        }
        return true;
    }

    @Override
    public boolean onLongClick(View v) {
        if (list.size() > 1) {
            int position = (Integer) v.getTag();
            T cmd = list.remove(position);
            startDrag(v, cmd, position);
            notifyDataSetChanged();
        }
        return true;
    }

    public final void startDrag(View v, T cmd) {
        startDrag(v, cmd, -1);
    }

    protected abstract String getItemName(T item);

    private final void startDrag(View v, T cmd, int originatingLine) {
        String name = getItemName(cmd);
        v.startDrag(ClipData.newPlainText(name, name), new View.DragShadowBuilder(v),
                new Object[] {
                        cmd, originatingLine
                }, 0);
        dragMinInsertPos = -1;
        dragMaxInsertPos = list.size();
    }

    String getActionStr(int action) {
        switch (action) {
            case DragEvent.ACTION_DRAG_ENDED:
                return "Ended";
            case DragEvent.ACTION_DRAG_STARTED:
                return "Started";
            case DragEvent.ACTION_DRAG_ENTERED:
                return "Entered";
            case DragEvent.ACTION_DRAG_EXITED:
                return "Exited";
            case DragEvent.ACTION_DRAG_LOCATION:
                return "Location";
            case DragEvent.ACTION_DROP:
                return "Drop";
        }
        return "???";
    }

    private View touchView;

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        T type = (T)v.getTag();
        switch (event.getAction()) {
            case MotionEvent.ACTION_UP:
                if (touchView == v) {
                    list.add(type);
                    notifyDataSetChanged();
                }
                break;
            case MotionEvent.ACTION_MOVE:
                if (list.size() > 0 && (event.getDownTime() - SystemClock.uptimeMillis()) > 500) {
                    startDrag(v, type);
                    touchView = null;
                }
                break;
            case MotionEvent.ACTION_DOWN:
                touchView = v;
                break;
        }

        return true;
    }

    public void addDraggable(View v, T type) {
        v.setTag(type);
        v.setOnTouchListener(this);
    }
}

File: AndroidGameLibrary2/src/main/java/cc/lib/android/DragAndDropAdapter.kt
package cc.lib.android

import android.content.ClipData
import android.os.SystemClock
import android.util.Log
import android.view.DragEvent
import android.view.MotionEvent
import android.view.View
import android.view.View.DragShadowBuilder
import android.view.View.OnDragListener
import android.view.View.OnTouchListener
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ListView
import android.widget.TextView
import java.util.Arrays

/**
 * Adapter to support drag and drop of list items within the list.
 * Also supports ability to drop external items into the list as droppable positions.
 *
 * Example:
 *
 * Say we have a view with ListView and some buttons like this:
 *
 * +-----------------------------+
 * |  +-----------------------+  |
 * |  |                       |  |
 * |  |       ListView        |  |
 * |  |                       |  |
 * |  |                       |  |
 * |  |                       |  |
 * |  |                       |  |
 * |  |                       |  |
 * |  |                       |  |
 * |  |                       |  |
 * |  +-----------------------+  |
 * |                             |
 * |  +-------+-------+-------+  |
 * |  |   A   |   B   |   C   |  |
 * |  +-------+-------+-------+  |
 * +-----------------------------+
 *
 * enum MyEnum {
 * A,B,C
 * }
 *
 * ListView listView;
 * Button buttonA,buttonB,buttonC;
 *
 * DragAndDropAdapter<MyEnum> adapter = new DrapAndDropAdapter<MyEnum>(listView) {
 * void populateItem(MyEnum e, ViewGroup container) {
 * TextView tv = new TextView(context);
 * tv.setText(e.name());
 * container.removeAllViews();
 * container.add(tv);
 * }
 *
 * String getItemName(MyEnum e) {
 * return e.name();
 * }
 * });
 *
 * adapter.addDraggable(buttonA, MyEnum.A);
 * adapter.addDraggable(buttonB, MyEnum.B);
 * adapter.addDraggable(buttonC, MyEnum.C);
 *
 * list.setAdapter(adapter);
 *
 * List<MyEnum> list = adapter.getList();
 *
 * @param <T>
</T></MyEnum></MyEnum></MyEnum> */
abstract class DragAndDropAdapter<T> @JvmOverloads constructor(
	private val listView: ListView, // TODO: Use dataset observer instead
	private val list: MutableList<T> = ArrayList()
) : BaseAdapter(), OnDragListener, View.OnLongClickListener, OnTouchListener {
	private var dragInsertPos = -1
	private var dragMinInsertPos = 0
	private var dragMaxInsertPos = 0
	private var dropped = false

	/**
	 * Populate a list item. Add views to the container. A non-empty container
	 * can have its views reused as appropriate.
	 *
	 * @param data
	 * @param container
	 */
	protected abstract fun populateItem(data: T, container: ViewGroup?, position: Int)

	/**
	 *
	 *
	 * @param item
	 * @return
	 */
	protected abstract fun getItemName(item: T): String

	constructor(listView: ListView, vararg items: T) : this(listView, ArrayList<T>(Arrays.asList<T>(*items)))

	override fun notifyDataSetChanged() {
		super.notifyDataSetChanged()
	}

	fun getList(): List<T> {
		return list
	}

	override fun getCount(): Int {
		val s = list.size
		return list.size
	}

	override fun getItem(position: Int): T {
		return list[position]
	}

	override fun getItemId(position: Int): Long {
		return position.toLong()
	}

	override fun getView(position: Int, convertView: View, parent: ViewGroup): View {
		var convertView = convertView
		if (convertView == null) {
			convertView = View.inflate(listView.context, R.layout.draganddrop_list_item, null)
		}
		convertView.setOnDragListener(this)
		convertView.setOnLongClickListener(this)
		convertView.tag = position
		val divTop = convertView.findViewById<View>(R.id.divider_top)
		val divBot = convertView.findViewById<View>(R.id.divider_bottom)
		divTop.visibility = View.GONE
		divBot.visibility = View.GONE
		if (dragInsertPos >= 0) {
			if (dragInsertPos == position) {
				divTop.visibility = View.VISIBLE
			} else if (dragInsertPos == position + 1) {
				divBot.visibility = View.VISIBLE
			}
		}
		val data = list[position]
		val tvLineNum = convertView.findViewById<TextView>(R.id.tvLineNum)
		tvLineNum.text = (position + 1).toString()
		onLineNumTextViewUpdated(position, tvLineNum)
		val v = convertView.findViewById<View>(R.id.ibDelete)
		v.visibility = View.VISIBLE
		v.tag = position
		v.setOnClickListener { v1: View ->
			val position1 = v1.tag as Int
			val data1 = list.removeAt(position1)
			notifyDataSetChanged()
		}
		onDeleteButtonUpdated(position, v)
		val container = convertView.findViewById<ViewGroup>(R.id.container)
		//container.removeAllViews();
		populateItem(data, container, position)
		return convertView
	}

	fun onLineNumTextViewUpdated(position: Int, textView: TextView?) {}
	fun onDeleteButtonUpdated(position: Int, button: View?) {}
	override fun onDrag(v: View, event: DragEvent): Boolean {
		val dragInsertPosIn = dragInsertPos
		val state = event.localState as Array<Any> ?: return false
		val data = state[0] as T
		val originatingLine = state[1] as Int
		val position = v.tag as Int
		val obj = "RowItem[$position]"
		when (event.action) {
			DragEvent.ACTION_DROP -> {
				Log.d(TAG, "dragInsertPos=$dragInsertPos")
				if (dragInsertPos < 0 || list.size == 0) {
					list.add(data)
				} else {
					list.add(dragInsertPos, data)
				}
				notifyDataSetChanged()
				listView.setSelection(dragInsertPos)
				dragInsertPos = -1
				dropped = true
			}

			DragEvent.ACTION_DRAG_EXITED -> {}
			DragEvent.ACTION_DRAG_LOCATION, DragEvent.ACTION_DRAG_ENTERED -> {
				val mp = (v.height / 2).toFloat()
				val mpd = event.y // + dragging.getHeight();
				val first = listView.firstVisiblePosition
				val last = listView.lastVisiblePosition

				//if (position == first) {
				//    Log.d("PLV", "A");
				//    dragInsertPos = position;
				//} else if (position == last) {
				//    Log.d("PLV", "B");
				//    dragInsertPos = position+1;
				//} else
				dragInsertPos = if (mp > mpd) {
					Log.d(TAG, "C mp=$mp mpd=$mpd")
					position
				} else {
					Log.d(TAG, "D")
					position + 1
				}
				dragInsertPos = dragInsertPos.coerceIn(dragMinInsertPos, dragMaxInsertPos)
				notifyDataSetChanged()
				//listView.setSelection(dragInsertPos);
				if (dragInsertPosIn != dragInsertPos) {
					val scrollDist = v.height
					val midpt = (last + first + 1) / 2
					val maxSp = 500f
					val speed = Math.round(maxSp - Math.abs(maxSp * 2 * (dragInsertPos - midpt) / (last - first)))
					//0.5f * (dragInsertPos - first - midpt));
					Log.d(TAG, "speed=$speed")
					if (dragInsertPos < midpt) {
						listView.smoothScrollBy(-scrollDist, speed)
					} else {
						listView.smoothScrollBy(scrollDist, speed)
					}
				}
				if (dragInsertPos == 0) listView.smoothScrollToPositionFromTop(0, 0, 0)
				if (dragInsertPos <= first + 1) {
					//listView.smoothScrollBy(-scrollDist, 500);
				} else if (dragInsertPos >= last - 1) {
					//  smoothScrollBy(scrollDist, 500);
				}
			}

			DragEvent.ACTION_DRAG_STARTED -> dropped = false
			DragEvent.ACTION_DRAG_ENDED -> if (!dropped) {
				// put the item back in
				if (originatingLine >= 0) list.add(originatingLine, data) else list.add(data)
				notifyDataSetChanged()
				dropped = true
			}
		}
		//if (dragInsertPos != dragInsertPosIn)
		run {
			Log.d(
				TAG,
				"v=" + obj + " action=" + getActionStr(event.action) + " pos=" + dragInsertPos + " origLine: " + originatingLine
			)
		}
		return true
	}

	override fun onLongClick(v: View): Boolean {
		if (list.size > 1) {
			listView.transcriptMode = ListView.TRANSCRIPT_MODE_DISABLED
			val position = v.tag as Int
			val data = list.removeAt(position)
			startDrag(v, data, position)
			notifyDataSetChanged()
		}
		return true
	}

	private fun startDrag(v: View, data: T, originatingLine: Int = -1) {
		val name = getItemName(data)
		v.startDrag(
			ClipData.newPlainText(name, name), DragShadowBuilder(v), arrayOf(
				data, originatingLine
			), 0
		)
		dragMinInsertPos = -1
		dragMaxInsertPos = list.size
	}

	private var touchView: View? = null

	/**
	 *
	 * @param listView
	 */
	init {
		listView.adapter = this
	}

	override fun onTouch(v: View, event: MotionEvent): Boolean {
		val type = v.tag as T
		when (event.action) {
			MotionEvent.ACTION_UP -> if (touchView === v) {
				listView.transcriptMode = ListView.TRANSCRIPT_MODE_ALWAYS_SCROLL
				list.add(type)
				listView.setSelection(list.size - 1)
				notifyDataSetChanged()
			}

			MotionEvent.ACTION_MOVE -> {
				val delta = SystemClock.uptimeMillis() - event.downTime
				//Log.d(TAG, "delta -= " + delta);
				if (list.size > 0 && delta > 500) {
					startDrag(v, type)
					touchView = null
				}
			}

			MotionEvent.ACTION_DOWN -> {
				listView.transcriptMode = ListView.TRANSCRIPT_MODE_DISABLED
				touchView = v
			}
		}
		return true
	}

	fun addDraggable(v: View, type: T) {
		v.tag = type
		v.setOnTouchListener(this)
	}

	companion object {
		val TAG = DragAndDropAdapter::class.java.simpleName
		private fun getActionStr(action: Int): String {
			when (action) {
				DragEvent.ACTION_DRAG_ENDED -> return "Ended"
				DragEvent.ACTION_DRAG_STARTED -> return "Started"
				DragEvent.ACTION_DRAG_ENTERED -> return "Entered"
				DragEvent.ACTION_DRAG_EXITED -> return "Exited"
				DragEvent.ACTION_DRAG_LOCATION -> return "Location"
				DragEvent.ACTION_DROP -> return "Drop"
			}
			return "???"
		}
	}
}
